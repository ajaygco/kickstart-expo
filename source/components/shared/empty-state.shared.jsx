// Modules
import React from "react";
import { Text, View } from "react-native";

// Shared
import { If } from "components/shared/if.shared";

// Component: Presentation
export const EmptyStateUi = ({ title, message, illustration }) => {
  const shouldShowIllustration = React.isValidElement(illustration);

  return (
    <View className="w-full items-center justify-center p-16">
      <If condition={shouldShowIllustration}>
        <View className="w-64 h-64 flex items-center justify-center mb-8">
          {illustration}
        </View>
      </If>

      <If condition={title}>
        <Text
          allowFontScaling={false}
          className="text-xl text-gray-800 font-sans font-bold mb-2"
        >
          {title}
        </Text>
      </If>

      <If condition={message}>
        <Text
          allowFontScaling={false}
          className="text-base text-gray-600 text-center font-sans"
        >
          {message}
        </Text>
      </If>
    </View>
  );
};

// Component: Logic
export const EmptyStateLogic = (props) => {
  return <EmptyStateUi {...props} />;
};

// Component: Memo
export const EmptyState = React.memo((props) => {
  return <EmptyStateLogic {...props} />;
});
