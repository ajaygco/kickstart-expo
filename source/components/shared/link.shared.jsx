// Modules
import React from "react";
import { Linking, Pressable } from "react-native";

// Helpers
import { Alert } from "helpers/alert.helper";
import { captureException } from "helpers/analytics.helper";

// Component: Presentation
export const LinkUi = ({ onPress, children, ...props }) => {
  return (
    <Pressable hitSlop={12} onPress={onPress} {...props}>
      {children}
    </Pressable>
  );
};

// Component: Logic
export const LinkLogic = ({ to, ...props }) => {
  // On Press
  const onPress = async () => {
    try {
      await Linking.openURL(to);
    } catch (error) {
      captureException(error);

      Alert.error({
        title: "Error",
        message: "Couldn't open link",
      });
    }
  };

  return <LinkUi onPress={onPress} {...props} />;
};

// Component: Memo
export const Link = React.memo((props) => {
  return <LinkLogic {...props} />;
});
