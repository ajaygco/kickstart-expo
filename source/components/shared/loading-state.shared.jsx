// Modules
import React from "react";
import { ActivityIndicator, View } from "react-native";

// Component: Presentation
export const LoadingStateUi = (props) => {
  return (
    <View className="items-center justify-center">
      <ActivityIndicator {...props} />
    </View>
  );
};

// Component: Logic
export const LoadingStateLogic = (props) => {
  return <LoadingStateUi {...props} />;
};

// Component: Memo
export const LoadingState = React.memo((props) => {
  return <LoadingStateLogic {...props} />;
});
