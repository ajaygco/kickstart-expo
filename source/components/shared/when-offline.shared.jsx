// Modules
import React from "react";
import { useIsConnected } from "react-native-offline";

// Component: Presentation
export const WhenOfflineUi = ({ isConnected, children }) => {
  if (isConnected) {
    return;
  }

  return children;
};

// Component: Logic
export const WhenOffline = ({ ...props }) => {
  // Hooks
  const isDeviceConnected = useIsConnected();

  return <WhenOfflineUi isConnected={isDeviceConnected} {...props} />;
};
