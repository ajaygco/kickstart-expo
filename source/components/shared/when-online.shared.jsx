// Modules
import React from "react";
import { useIsConnected } from "react-native-offline";

// Component: Presentation
export const WhenOnlineUi = ({ isConnected, children }) => {
  if (!isConnected) {
    return;
  }

  return children;
};

// Component: Logic
export const WhenOnline = (props) => {
  const isConnected = useIsConnected();

  return <WhenOnlineUi isConnected={isConnected} {...props} />;
};
