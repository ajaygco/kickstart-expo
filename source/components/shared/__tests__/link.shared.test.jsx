// bunx jest source/components/shared/__tests__/link.shared.test.jsx

import React from "react";
import { Text, Linking } from "react-native";
import { fireEvent, render } from "@testing-library/react-native";

import { Link } from "components/shared/link.shared";

jest.mock("helpers/analytics.helper", () => ({
  captureException: jest.fn(),
}));

jest.mock("react-native/Libraries/Linking/Linking", () => ({
  openURL: jest.fn(() => Promise.resolve(true)),
}));

describe("Link", () => {
  it("should render correctly", () => {
    const wrapper = render(
      <Link to="https://example.com" testID="element">
        <Text>Visit Website</Text>
      </Link>,
    );

    const element = wrapper.getByTestId("element");

    const text = wrapper.getByText("Visit Website");

    expect(text).toBeDefined();
    expect(element).toBeDefined();

    expect(wrapper).toMatchSnapshot();
  });

  it("should open the 'to' link when pressed", async () => {
    const wrapper = render(
      <Link to="https://example.com" testID="element">
        <Text>Visit Website</Text>
      </Link>,
    );

    const element = wrapper.getByTestId("element");

    expect(element).toBeDefined();

    fireEvent.press(element);
    expect(Linking.openURL).toHaveBeenCalledWith("https://example.com");

    expect(wrapper).toMatchSnapshot();
  });
});
