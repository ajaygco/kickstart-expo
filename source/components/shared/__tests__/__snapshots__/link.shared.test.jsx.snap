// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should open the 'to' link when pressed 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={12}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="element"
>
  <Text>
    Visit Website
  </Text>
</View>
`;

exports[`Link should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={12}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="element"
>
  <Text>
    Visit Website
  </Text>
</View>
`;
